pkgdb2/lib/model/__init__.py

```diff
@@ -1590,6 +1641,191 @@ def insert(cls, session, user, package, description):
         session.flush()
 
 
+class AdminAction(BASE):
+    """This associates a package with a particular collection.
+
+    Table -- admin_actions
+    """
+
+    __tablename__ = 'admin_actions'
+    id = sa.Column(sa.Integer, nullable=False, primary_key=True)
+    package_id = sa.Column(
+        sa.Integer,
+        sa.ForeignKey(
+            'Package.id', ondelete="CASCADE", onupdate="CASCADE"),
+        nullable=True)
+    collection_id = sa.Column(
+        sa.Integer,
+        sa.ForeignKey(
+            'Collection.id', ondelete="CASCADE", onupdate="CASCADE"),
+        nullable=False)
+    _status = sa.Column(
+        sa.String(50),
+        sa.ForeignKey('action_status.status', onupdate='CASCADE'),
+        name='status',
+        nullable=False,
+        index=True)
+    user = sa.Column(sa.Text, nullable=False, index=True)
+    action = sa.Column(sa.Text, nullable=False, index=True)
+    from_collection_id = sa.Column(
+        sa.Integer,
+        sa.ForeignKey(
+            'Collection.id', ondelete="CASCADE", onupdate="CASCADE"),
+        nullable=True)
+    info = sa.Column(sa.Text, nullable=True)
+    message = sa.Column(sa.Text, nullable=True)
+
+    date_created = sa.Column(sa.DateTime, nullable=False,
+                             default=datetime.datetime.utcnow)
+    date_change = sa.Column(sa.DateTime, nullable=False,
+                            default=datetime.datetime.utcnow,
+                            onupdate=sa.func.now())
+
+    __table_args__ = (
+        sa.UniqueConstraint(
+            'user', 'action', 'status', 'package_id', 'collection_id'),
+    )
+
+    package = relation(
+        "Package",
+        backref=backref("requests", order_by=collection_id)
+    )
+    collection = relation(
+        "Collection",
+        foreign_keys=[collection_id], remote_side=[Collection.id],
+    )
+    from_collection = relation(
+        "Collection",
+        foreign_keys=[from_collection_id], remote_side=[Collection.id],
+    )
+
+    @property
+    def info_data(self):
+        """ Return the dict stored as string in the database as an actual
+        dict object.
+
+        """
+        if self.info:
+            return ast.literal_eval(self.info)
```

Comment:
Whoah.  ``literal_eval``?  I see that the docstring claims it is 'safe', but I've only ever seen ``json`` used for situations like these.